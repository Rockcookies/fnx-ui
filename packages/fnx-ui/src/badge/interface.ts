import { ReactNode, HTMLAttributes } from 'react';

export type BadgeProps = BadgeComponentProps &
	Omit<HTMLAttributes<HTMLSpanElement>, 'color'>;

export interface BadgeComponentProps {
	count?: ReactNode;
	showZero?: boolean;
	dot?: boolean;
	max?: string | number;
	color?: string;
	offset?: [number | string, number | string];
}
