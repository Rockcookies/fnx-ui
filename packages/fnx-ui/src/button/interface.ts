import { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonType =
	| 'default'
	| 'primary'
	| 'success'
	| 'warning'
	| 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm' | 'xs';

export type ButtonShape = 'rect' | 'radius' | 'round';

export interface ButtonComponentProps {
	type?: ButtonType;
	size?: ButtonSize;
	icon?: ReactNode;
	iconPosition?: 'left' | 'right';
	loading?: boolean;
	loadingIcon?: ReactNode;
	loadingChildren?: ReactNode;
	disabled?: boolean;
	hairline?: boolean;
	plain?: boolean;
	shape?: ButtonShape;
	block?: boolean;
	color?: string;
	htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

export type ButtonProps = ButtonComponentProps &
	Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'>;
