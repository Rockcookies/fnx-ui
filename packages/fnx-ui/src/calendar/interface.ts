import { HTMLAttributes, ReactNode } from 'react';
import CalendarDay from './CalendarDay';
import CalendarMulti from './CalendarMulti';
import CalendarRange from './CalendarRange';

export type CalendarValue = Date | string | number;

export type CalendarDayType =
	| 'normal'
	| 'start'
	| 'start-end'
	| 'middle'
	| 'end'
	| 'selected'
	| 'multiple-middle'
	| 'multiple-selected'
	| 'disabled';

export interface CalendarDayComponentProps {
	color?: string;
	dayHeight?: string;
	date: Date;
	content?: ReactNode;
	type: CalendarDayType;
	topInfo?: ReactNode;
	bottomInfo?: ReactNode;
}

export type CalendarDayProps = CalendarDayComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'content'>;

export interface CalendarSlots {
	day?: (day: CalendarDayComponentProps) => ReactNode;
	subTitle?: boolean | ((date: Date) => ReactNode);
	monthMark?: boolean | ((date: Date) => ReactNode);
	monthTitle?: boolean | ((date: Date) => ReactNode);
	footer?: ReactNode;
}

export interface CalendarBaseProps {
	title?: ReactNode;
	color?: string;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	dayHeight?: number | string;
	readonly?: boolean;
	closeIcon?: ReactNode;
	firstDayOfWeek?: number;
	slots?: CalendarSlots;
	confirmText?: ReactNode;
	confirmDisabledText?: ReactNode;
}

export type CalendarElementProps = Omit<
	HTMLAttributes<HTMLDivElement>,
	'defaultValue' | 'onChange' | 'title'
>;

export interface CalendarContainerRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date[] | undefined;
}

export interface CalendarRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date | undefined;
}

export interface CalendarComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue;
	onChange?: (value: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: () => void;
}

export type CalendarProps = CalendarComponentProps & CalendarElementProps;

export type CalendarRangeRef = CalendarContainerRef;

export interface CalendarRangeComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	allowSameDay?: boolean;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarRangeProps = CalendarRangeComponentProps &
	CalendarElementProps;

export type CalendarMultiRef = CalendarRangeRef;

export interface CalendarMultiComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarMultiProps = CalendarMultiComponentProps &
	CalendarElementProps;

export interface CalendarExportExtra {
	Range: typeof CalendarRange;
	Multi: typeof CalendarMulti;
	Day: typeof CalendarDay;
}
