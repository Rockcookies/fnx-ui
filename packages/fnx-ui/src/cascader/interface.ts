import { HTMLAttributes, ReactNode } from 'react';

export type CascaderValue = string | number;

export interface CascaderOption {
	label?: ReactNode;
	value?: CascaderOption;
	loading?: boolean;
	disabled?: boolean;
	children?: CascaderOption[];
	// for custom filed names
	[key: string]: any;
}

export interface CascaderDataNames {
	label?: string | string[];
	value?: string | string[];
	loading?: string | string[];
	disabled?: string | string[];
	children?: string | string[];
}

export interface CascaderComponentProps<T = CascaderOption> {
	defaultValue?: CascaderValue[];
	data?: T[];
	dataNames?: CascaderDataNames;
	onLoadData?: (options: T[]) => void;
	onConfirm?: (value: CascaderValue[], options: T[]) => void;
	onCancel?: () => void;
	onChange?: (value: CascaderValue[], options: T[]) => void;
	title?: ReactNode;
	placeholder?: ReactNode;
	swipeable?: boolean;
	closeIcon?: ReactNode;
	activeColor?: string;
}

export type CascaderProps<T = CascaderOption> = CascaderComponentProps<T> &
	Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'title' | 'onChange'>;
