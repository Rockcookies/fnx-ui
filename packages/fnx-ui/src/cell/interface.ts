import { HTMLAttributes, ReactNode } from 'react';
import CellGroup from './CellGroup';

export type CellItemsAlign = 'top' | 'middle' | 'bottom';

export interface CellComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	content?: ReactNode;
	contentProps?: HTMLAttributes<HTMLDivElement>;
	description?: ReactNode;
	descriptionProps?: HTMLAttributes<HTMLDivElement>;
	leftIcon?: ReactNode;
	leftIconProps?: HTMLAttributes<HTMLSpanElement>;
	rightIcon?: ReactNode;
	rightIconProps?: HTMLAttributes<HTMLSpanElement>;
	border?: boolean;
	itemsAlign?: CellItemsAlign;
	clickable?: boolean;
	size?: 'md' | 'lg';
}

export type CellProps = CellComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellGroupComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	border?: boolean;
	insert?: boolean;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
}

export type CellGroupProps = CellGroupComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellExportExtra {
	Group: typeof CellGroup;
}
