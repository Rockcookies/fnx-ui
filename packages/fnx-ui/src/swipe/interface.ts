import { HTMLAttributes, ReactNode } from 'react';
import SwipeItem from './SwipeItem';

export interface SwipeComponentProps {
	width?: number;
	height?: number;
	onActiveIndexChange?: (current: number) => void;
	autoplay?: number;
	duration?: number;
	defaultActiveIndex?: number;
	loop?: boolean;
	vertical?: boolean;
	touchable?: boolean;
	indicator?: ReactNode;
	stopPropagation?: boolean;
}

export type SwipeProps = SwipeComponentProps & HTMLAttributes<HTMLDivElement>;

export type SwipeItemProps = HTMLAttributes<HTMLDivElement>;

export interface SwipeRef {
	root: HTMLElement | null;
	wrapper: HTMLElement | null;
	reset: () => void;
	next: () => void;
	prev: () => void;
	swipeTo: (index: number, options?: { animation?: boolean }) => void;
}

export interface SwipeExportExtra {
	Item: typeof SwipeItem;
}
