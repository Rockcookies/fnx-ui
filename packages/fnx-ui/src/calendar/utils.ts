import { createBEM } from '../utils/namespace';

export const _bem = createBEM('fnx-calendar');

export type CalendarMode = 'single' | 'multiple' | 'range';

export interface CalendarMonthRef {
	root: HTMLDivElement | null;
	getHeight(): number;
}

export function compareMonth(d1: Date, d2: Date) {
	const year1 = d1.getFullYear();
	const year2 = d2.getFullYear();

	if (year1 === year2) {
		const month1 = d1.getMonth();
		const month2 = d2.getMonth();

		return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
	}

	return year1 > year2 ? 1 : -1;
}

export function compareDate(d1: Date, d2: Date) {
	const res = compareMonth(d1, d2);

	if (res === 0) {
		const date1 = d1.getDate();
		const date2 = d2.getDate();

		return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
	}

	return res;
}

export function isCalendarDateEquals(d1: Date, d2: Date) {
	return (
		d1.getFullYear() === d2.getFullYear() &&
		d1.getMonth() === d2.getMonth() &&
		d1.getDate() === d2.getDate()
	);
}

export const cloneDate = (date: Date) => new Date(date);

export const getPrevDay = (date: Date) => getDayByOffset(date, -1);
export const getNextDay = (date: Date) => getDayByOffset(date, 1);
export const getToday = () => {
	const today = new Date();
	today.setHours(0, 0, 0, 0);
	return today;
};

export function getDayByOffset(date: Date, offset: number) {
	const cloned = cloneDate(date);
	cloned.setDate(cloned.getDate() + offset);
	return cloned;
}

export function calcDateNum(d1: Date, d2: Date) {
	const day1 = d1.getTime();
	const day2 = d2.getTime();
	return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}
