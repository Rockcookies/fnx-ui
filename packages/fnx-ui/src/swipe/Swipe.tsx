import React, {
	CSSProperties,
	forwardRef,
	isValidElement,
	ReactNode,
	useCallback,
	useEffect,
	useImperativeHandle,
	useMemo,
	useRef,
} from 'react';
import configComponentProps from '../hooks/config-component-props';
import useMergedPropRef from '../hooks/use-merged-prop-ref';
import useWindowSize from '../hooks/use-window-size';
import useOnPopupOpen from '../popup/hooks/use-on-popup-open';
import {
	bindEvent,
	listenDocumentVisibilityChange,
	preventDefault,
} from '../utils/dom/event';
import TouchHelper from '../utils/dom/touch-helper';
import { noop } from '../utils/misc';
import { classnames, createBEM } from '../utils/namespace';
import { doubleRaf } from '../utils/raf';
import { toElementArray } from '../utils/react';
import useSwipe from './hooks/use-swipe';
import { SwipeComponentProps, SwipeProps, SwipeRef } from './interface';

const NS = 'fnx-swipe';
const bem = createBEM(NS);

const useProps = configComponentProps<Required<SwipeComponentProps>>({
	height: 0,
	width: 0,
	autoplay: 3000,
	duration: 500,
	defaultActiveIndex: 0,
	loop: true,
	vertical: false,
	touchable: true,
	indicator: null,
	stopPropagation: true,
	onActiveIndexChange: noop,
});

const Swipe = forwardRef<SwipeRef, SwipeProps>((_props, ref) => {
	const [props, { className, children: _children, ...restProps }] =
		useProps(_props);

	const { autoplay, duration, vertical, touchable, indicator } = props;

	const propsRef = useMergedPropRef<Required<SwipeComponentProps>>(props);

	const rootRef = useRef<HTMLDivElement>(null);
	const wrapperRef = useRef<HTMLDivElement>(null);

	const windowSize = useWindowSize();

	const children = toElementArray(_children);
	const swipeLength = children.filter((v) => React.isValidElement(v)).length;

	const activeIndexInitializedRef = useRef(false);

	const {
		stateRef,
		wrapperOffset,
		setWrapperOffset,
		swipeSize,
		swipeOffset,
		activeIndex,
		setActiveIndex,
		swiping,
		move,
		next,
		prev,
		stopTransition,
		stopAutoplay,
		startAutoplay,
		reload,
	} = useSwipe({
		propsRef,
		swipeLength,
		rootRef,
	});

	const reset = useCallback(
		(action?: () => void) => {
			stopTransition();
			stopAutoplay();
			action && action();
			startAutoplay();
		},
		[startAutoplay, stopAutoplay, stopTransition],
	);

	const swipeRef = useMemo<SwipeRef>(() => {
		return {
			root: rootRef.current || null,
			wrapper: wrapperRef.current || null,
			reset: () => {
				reset(reload);
			},
			next: () => {
				reset(next);
			},
			prev: () => {
				reset(prev);
			},
			swipeTo: (idx, options) => {
				reset(() => {
					const { swipeSize } = stateRef.current;
					setActiveIndex(idx);
					move(-idx * swipeSize, {
						swiping: options && options.animation,
					});
				});
			},
		};
	}, [move, next, prev, reload, reset, setActiveIndex, stateRef]);

	useImperativeHandle<SwipeRef, SwipeRef>(ref, () => swipeRef, [swipeRef]);

	const itemStyles = useMemo<CSSProperties[]>(() => {
		const output: CSSProperties[] = [];

		for (let i = 0; i < swipeLength; i++) {
			const formattedStyle: CSSProperties = {};

			if (swipeSize > 0 && swipeOffset) {
				if (i === 0 && swipeOffset === 'first') {
					formattedStyle.transform = `translate${
						vertical ? 'Y' : 'X'
					}(${swipeLength * swipeSize}px)`;
				}

				if (i === swipeLength - 1 && swipeOffset === 'last') {
					formattedStyle.transform = `translate${
						vertical ? 'Y' : 'X'
					}(-${swipeLength * swipeSize}px)`;
				}
			}

			if (swipeSize > 0) {
				formattedStyle[
					vertical ? 'height' : 'width'
				] = `${swipeSize}px`;
			}

			output.push(formattedStyle);
		}

		return output;
	}, [swipeOffset, vertical, swipeSize, swipeLength]);

	useEffect(() => {
		if (autoplay > 0) {
			startAutoplay();

			return () => {
				stopAutoplay();
			};
		}
	}, [autoplay, startAutoplay, stopAutoplay]);

	useEffect(() => {
		if (activeIndex == null) {
			return;
		}

		if (activeIndexInitializedRef.current) {
			propsRef.current.onActiveIndexChange(activeIndex);
		} else {
			activeIndexInitializedRef.current = true;
		}
	}, [activeIndex, propsRef]);

	useEffect(() => {
		return listenDocumentVisibilityChange((visibility) => {
			if (visibility === 'hidden') {
				stopAutoplay();
			} else {
				startAutoplay();
			}
		});
	}, [startAutoplay, stopAutoplay]);

	useEffect(() => {
		reset(reload);
	}, [swipeLength, windowSize, startAutoplay, reload, stopAutoplay, reset]);

	useOnPopupOpen(() => doubleRaf(() => reset(reload)));

	useEffect(() => {
		const node = wrapperRef.current;

		if (!node || touchable === false) {
			return;
		}

		const touch = new TouchHelper();
		let touchStartTime = 0;
		let delta = 0;
		let touchWrapperOffset = 0;
		let touching = false;

		const onTouchStart = (event: TouchEvent) => {
			touch.start(event);
			touchStartTime = Date.now();
			touching = false;
		};

		const onTouchMove = (event: TouchEvent) => {
			touch.move(event);

			const { vertical, stopPropagation, loop } = propsRef.current;

			const touchData = touch.touchData;

			if (!touching) {
				const direction = vertical ? 'vertical' : 'horizontal';

				if (touchData.direction === direction) {
					stopTransition();
					stopAutoplay();
					reload();

					touching = true;
					touchWrapperOffset = stateRef.current.wrapperOffset || 0;
				}
			} else {
				delta = touch.isVertical()
					? touch.touchData.deltaY
					: touch.touchData.deltaX;

				move(touchWrapperOffset + delta, { critical: loop });
			}

			const shouldPrevent =
				touching && touchData.offsetY > touchData.offsetX === vertical;

			if (shouldPrevent) {
				preventDefault(event, stopPropagation);
			}
		};

		const restore = () => {
			const duration = Date.now() - touchStartTime;
			const speed = delta / duration;

			const {
				activeIndex = 0,
				swipeSize,
				swipeLength,
			} = stateRef.current;
			const { loop } = propsRef.current;

			const quickSwipe = Math.abs(speed) > 0.25;

			const shouldSwipe =
				(quickSwipe || Math.abs(delta) > swipeSize / 2) &&
				swipeSize > 0 &&
				swipeLength > 1;

			if (!shouldSwipe) {
				move(-activeIndex * swipeSize, {
					swiping: true,
				});

				return;
			}

			if (quickSwipe) {
				if (loop) {
					delta < 0 ? next() : prev();
				} else if (delta < 0 && activeIndex < swipeLength - 1) {
					next();
				} else if (delta > 0 && activeIndex > 0) {
					prev();
				} else {
					move(-activeIndex * swipeSize, {
						swiping: true,
					});
				}

				return;
			}

			touchWrapperOffset = touchWrapperOffset + delta;

			if (touchWrapperOffset > 0) {
				setActiveIndex(loop ? swipeLength - 1 : 0);
				move(loop ? swipeSize : 0, {
					swiping: true,
					critical: loop,
				});
			} else if (-touchWrapperOffset > swipeSize * (swipeLength - 1)) {
				setActiveIndex(loop ? 0 : swipeLength - 1);
				move(-swipeSize * (loop ? swipeLength : swipeLength - 1), {
					swiping: true,
					critical: loop,
				});
			} else {
				const nextActiveIndex = Math.ceil(
					-touchWrapperOffset / swipeSize,
				);
				setActiveIndex(nextActiveIndex);
				move(-nextActiveIndex - swipeSize, {
					swiping: true,
					critical: loop,
				});
			}
		};

		const onTouchEnd = () => {
			if (!touching) {
				return;
			}

			restore();

			touching = false;
			startAutoplay();
		};

		node.addEventListener('touchstart', onTouchStart, false);
		bindEvent(node, 'touchmove', onTouchMove, { passive: false });
		node.addEventListener('touchend', onTouchEnd, false);
		node.addEventListener('touchcancel', onTouchEnd, false);

		return () => {
			node.removeEventListener('touchstart', onTouchStart);
			node.removeEventListener('touchmove', onTouchMove);
			node.removeEventListener('touchend', onTouchEnd);
			node.removeEventListener('touchcancel', onTouchEnd);
		};
	}, [
		move,
		next,
		prev,
		propsRef,
		reload,
		setActiveIndex,
		setWrapperOffset,
		startAutoplay,
		stateRef,
		stopAutoplay,
		stopTransition,
		touchable,
	]);

	const getWrapperStyle = (): CSSProperties => {
		const formattedStyle: CSSProperties = {};

		formattedStyle.transform = `translate${
			vertical ? 'Y' : 'X'
		}(${wrapperOffset}px)`;

		if (swipeSize > 0) {
			formattedStyle[vertical ? 'height' : 'width'] = `${
				swipeSize * swipeLength
			}px`;
		}

		if (swiping) {
			formattedStyle.transitionDuration = `${duration}ms`;
		}

		return formattedStyle;
	};

	const renderIndicators = (): ReactNode => {
		if (indicator === false || swipeLength <= 0) {
			return;
		}

		if (indicator == null || indicator === true) {
			return (
				<div className={bem('indicators')}>
					{new Array(swipeLength).fill(0).map((_, index) => (
						<i
							className={bem('indicator', {
								active: index === (activeIndex || 0),
							})}
							key={index}
						/>
					))}
				</div>
			);
		}

		return indicator;
	};

	return (
		<div
			ref={rootRef}
			className={classnames(bem({ vertical }), className)}
			{...restProps}
		>
			<div
				className={bem('wrapper')}
				style={getWrapperStyle()}
				ref={wrapperRef}
			>
				{children.map((item, idx) => {
					if (isValidElement<any>(item)) {
						const itemStyle = item.props.style;
						const itemKey = item.key;

						let mergedStyle: CSSProperties = {
							...itemStyles[idx],
						};

						if (typeof itemStyle === 'object') {
							mergedStyle = { ...mergedStyle, ...itemStyle };
						}

						return React.cloneElement(item, {
							...item.props,
							style: mergedStyle,
							key: itemKey != null ? itemKey : idx,
						});
					}

					return item;
				})}
			</div>

			{renderIndicators()}
		</div>
	);
});

Swipe.displayName = 'Swipe';

export default Swipe;
