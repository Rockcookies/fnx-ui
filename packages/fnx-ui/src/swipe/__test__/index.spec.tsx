import { act, render } from '@testing-library/react';
import React from 'react';
import TestsDOM from '../../../test/dom';
import TestsEvent from '../../../test/events';
import TestsReact from '../../../test/react';
import { ElementRect } from '../../utils/interface';
import Swipe from '../index';
import { SwipeProps, SwipeRef } from '../interface';

describe('<Swipe/>', () => {
	let eachClears: Array<() => void> = [];

	const mockGetBoundingClientRect = (
		rect: Partial<ElementRect>,
	): (() => void) => {
		const originMethod = Element.prototype.getBoundingClientRect;

		Element.prototype.getBoundingClientRect = jest.fn(
			() =>
				({
					top: 0,
					left: 0,
					right: 0,
					bottom: 0,
					width: 0,
					height: 0,
					...rect,
				}) as any,
		);

		return function () {
			Element.prototype.getBoundingClientRect = originMethod;
		};
	};

	const mockPageHidden = (visibility: 'hidden' | 'visible') => {
		Object.defineProperty(document, 'visibilityState', {
			get: () => visibility,
		});

		document.dispatchEvent(new Event('visibilitychange'));
	};

	const renderDemo = (props: SwipeProps = {}) => {
		const swipeRef = React.createRef<SwipeRef>();

		const { container, getByTestId } = render(
			<Swipe
				ref={swipeRef}
				vertical={false}
				autoplay={0}
				touchable={true}
				loop={true}
				data-testid="swipe"
				{...props}
			>
				<Swipe.Item>1</Swipe.Item>
				<Swipe.Item>2</Swipe.Item>
				<Swipe.Item>3</Swipe.Item>
			</Swipe>,
		);

		return { container, getByTestId, swipeRef };
	};

	beforeEach(() => {
		jest.useFakeTimers();

		const clear = mockGetBoundingClientRect({
			width: 100,
			height: 100,
		});

		eachClears.push(clear);
	});

	afterEach(() => {
		eachClears.forEach((clear) => clear());
		eachClears = [];

		jest.useRealTimers();
	});

	it('should swipe to specific swipe after calling the swipeTo method', () => {
		const onActiveIndexChange = jest.fn();

		const { swipeRef } = renderDemo({
			onActiveIndexChange,
		});

		act(() => {
			TestsReact.mustGetRef(swipeRef).swipeTo(2, { animation: true });
		});

		expect(onActiveIndexChange).toHaveBeenCalledWith(2);

		act(() => {
			TestsReact.mustGetRef(swipeRef).swipeTo(1, { animation: false });
		});

		expect(onActiveIndexChange).toHaveBeenCalledWith(1);
	});

	it('should swipe to next swipe after calling next method', () => {
		const onActiveIndexChange = jest.fn();

		const { swipeRef } = renderDemo({
			onActiveIndexChange,
		});

		act(() => {
			TestsReact.mustGetRef(swipeRef).next();
		});

		expect(onActiveIndexChange).toHaveBeenCalledWith(1);
	});

	it('should swipe to prev swipe after calling prev method', () => {
		const onActiveIndexChange = jest.fn();

		const { swipeRef } = renderDemo({
			onActiveIndexChange,
		});

		act(() => {
			TestsReact.mustGetRef(swipeRef).prev();
		});

		expect(onActiveIndexChange).toHaveBeenCalledWith(2);
	});

	it('should render initial swipe correctly', async () => {
		let demo = renderDemo();
		expect(demo.container).toMatchSnapshot();

		demo = renderDemo({ defaultActiveIndex: 2 });
		expect(demo.container).toMatchSnapshot();
	});

	it('should render vertical swipe when using vertical prop', async () => {
		const onActiveIndexChange = jest.fn();

		const { container } = renderDemo({
			vertical: true,
			onActiveIndexChange,
		});

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-swipe__wrapper',
		);

		await TestsEvent.triggerDrag(track, [0, -100]);

		expect(onActiveIndexChange).toHaveBeenCalledWith(1);
	});

	it('should not allow to drag swipe when touchable is false', async () => {
		const onActiveIndexChange = jest.fn();

		const { container } = renderDemo({
			touchable: false,
			onActiveIndexChange,
		});

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-swipe__wrapper',
		);

		await TestsEvent.triggerDrag(track, [100, 0]);

		expect(container).toMatchSnapshot();
		expect(onActiveIndexChange).toHaveBeenCalledTimes(0);
	});

	it('should render swipe looped when using loop prop', async () => {
		const onActiveIndexChange = jest.fn();

		const { container } = renderDemo({
			onActiveIndexChange,
		});

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-swipe__wrapper',
		);

		await TestsEvent.triggerDrag(track, [-100, 0]);
		expect(onActiveIndexChange).toHaveBeenLastCalledWith(1);

		await TestsEvent.triggerDrag(track, [-100, 0]);
		expect(onActiveIndexChange).toHaveBeenLastCalledWith(2);

		await TestsEvent.triggerDrag(track, [-100, 0]);
		expect(onActiveIndexChange).toHaveBeenLastCalledWith(0);

		await TestsEvent.triggerDrag(track, [-100, 0]);
		expect(onActiveIndexChange).toHaveBeenLastCalledWith(1);

		await TestsEvent.triggerDrag(track, [100, 0]);
		expect(onActiveIndexChange).toHaveBeenLastCalledWith(0);

		await TestsEvent.triggerDrag(track, [100, 0]);
		expect(onActiveIndexChange).toHaveBeenLastCalledWith(2);

		await TestsEvent.triggerDrag(track, [100, 0]);
		expect(onActiveIndexChange).toHaveBeenLastCalledWith(1);
	});

	it('should pause auto play when page hidden', async () => {
		const onActiveIndexChange = jest.fn();

		renderDemo({
			autoplay: 1,
			onActiveIndexChange,
		});

		act(() => {
			mockPageHidden('hidden');
		});

		act(() => {
			jest.advanceTimersByTime(50);
		});

		expect(onActiveIndexChange).toHaveBeenCalledTimes(0);
	});
});
