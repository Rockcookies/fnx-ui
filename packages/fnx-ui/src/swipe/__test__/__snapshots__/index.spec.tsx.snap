// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Swipe/> should not allow to drag swipe when touchable is false 1`] = `
<div>
  <div
    class="fnx-swipe"
    data-testid="swipe"
  >
    <div
      class="fnx-swipe__wrapper"
      style="transform: translateX(0px); width: 300px;"
    >
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        1
      </div>
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        2
      </div>
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        3
      </div>
    </div>
    <div
      class="fnx-swipe__indicators"
    >
      <i
        class="fnx-swipe__indicator fnx-swipe__indicator--active"
      />
      <i
        class="fnx-swipe__indicator"
      />
      <i
        class="fnx-swipe__indicator"
      />
    </div>
  </div>
</div>
`;

exports[`<Swipe/> should render initial swipe correctly 1`] = `
<div>
  <div
    class="fnx-swipe"
    data-testid="swipe"
  >
    <div
      class="fnx-swipe__wrapper"
      style="transform: translateX(0px); width: 300px;"
    >
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        1
      </div>
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        2
      </div>
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        3
      </div>
    </div>
    <div
      class="fnx-swipe__indicators"
    >
      <i
        class="fnx-swipe__indicator fnx-swipe__indicator--active"
      />
      <i
        class="fnx-swipe__indicator"
      />
      <i
        class="fnx-swipe__indicator"
      />
    </div>
  </div>
</div>
`;

exports[`<Swipe/> should render initial swipe correctly 2`] = `
<div>
  <div
    class="fnx-swipe"
    data-testid="swipe"
  >
    <div
      class="fnx-swipe__wrapper"
      style="transform: translateX(-200px); width: 300px;"
    >
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        1
      </div>
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        2
      </div>
      <div
        class="fnx-swipe-item"
        style="width: 100px;"
      >
        3
      </div>
    </div>
    <div
      class="fnx-swipe__indicators"
    >
      <i
        class="fnx-swipe__indicator"
      />
      <i
        class="fnx-swipe__indicator"
      />
      <i
        class="fnx-swipe__indicator fnx-swipe__indicator--active"
      />
    </div>
  </div>
</div>
`;
